File: py/algorithm/holt_winters_model.py
import numpy as np
import py.optimize.genetic_algorithm_optimizer as gao
import py.optimize.holt_winters_rmse_function as hwrf

class HoltWintersModel():

    def optimize(self, y, cycle_len, epoch_num=10000):
        boundaries = np.zeros((3, 2))
        for i in range(boundaries.shape[0]):
            boundaries[i][0] = 0.0
            boundaries[i][1] = 1.0
        optimizer = gao.GeneticAlgorithmOptimizer(boundaries, epoch_num)
        # initalize parameters that are going to be optimized
        params = np.zeros((1, 3))
        for i in range(params.shape[0]):
            params[i][0] = 0.001
        params = optimizer.optimize(hwrf.HoltWintersRmseFunction(cycle_len), params, y, y, False)
        return params

    def fittedValue(self, y, params, cycle_len, prediction_len):
        alpha = params[0]
        beta = params[1]
        gama = params[2]
        s = []
        b = []
        c = []
        yHat = []

        # Initialize s0
        sum = 0.0
        for i in range(cycle_len):
            sum = sum + y[i][0]
        s0 = sum / cycle_len
        s.append(s0)

        # Initialize b0
        sum = 0.0
        for i in range(cycle_len, 2 * cycle_len):
            sum = sum + (y[i][0] - y[i - cycle_len][0])
        b0 = sum / (cycle_len * cycle_len)
        b.append(b0)

        # Initialize c
        for i in range(cycle_len):
            ci = y[i][0] - s[0]
            c.append(ci)

        # Initialize yHat
        y0 = s[0] + b[0] + c[0]
        yHat.append(y0)

        # Estimate yHat
        for i in range(y.shape[0]):
            # estimate si, bi, ci, yHat
            s.append(alpha * (y[i][0] - c[i]) + (1 - alpha) * (s[i] + b[i]))
            b.append(beta * (s[i+1] - s[i]) + (1 - beta) * b[i])
            c.append(gama * (y[i][0] - s[i]) - b[i] + (1 - gama) * c[i])
            yHat.append(s[i + 1] + b[i + 1] + c[i + 1])
            if i == y.shape[0] - 1:
                for j in range(2, prediction_len + 2):
                    yHat.append(s[i + 1] + (b[i + 1] * j) + c[i + 1] + (j % prediction_len))

        yHat = yHat[1:]
        yHatMat = np.zeros((len(yHat), 1))
        for i in range(yHatMat.shape[0]):
            yHatMat[i][0] = yHat[i]
        return yHatMat

File: src/main/java/jm/app/algorithm/HoltWintersModel.java
package jm.app.algorithm;

import jm.app.algebra.AlgebraUtil;
import jm.app.algebra.Matrix;
import jm.app.optimize.GeneticAlgorithmOptimizer;
import jm.app.optimize.Optimizer;
import jm.app.optimize.TargetFunction;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.List;

public class HoltWintersModel {

    private final int DEFAULT_EPOCH_NUM = 20;

    public Matrix optimize(Matrix y, int cycleLen) {
        return optimize(y, cycleLen, DEFAULT_EPOCH_NUM);
    }

    public Matrix optimize(Matrix y, int cycleLen, int epochNum) {
        Matrix boundaries = new Matrix(3, 2);
        for (int i = 0; i < boundaries.getRowNum(); i++) {
            boundaries.setValue(i, 0, 0.0);
            boundaries.setValue(i, 1, 1.0);
        }
        Optimizer optimizer = new GeneticAlgorithmOptimizer(boundaries, epochNum);
        // initialize parameters that are going to be optimized
        Matrix params = new Matrix(1, 3);
        for (int i = 0; i < params.getRowNum(); i++) {
            params.setValue(i, 0, 0.001);
        }
        params = optimizer.optimize(new HoltWintersModelRmseTargetFunction(cycleLen), params, y, y, false);
        return params;
    }

    public Matrix fittedValue(Matrix y, Matrix params, int cycleLen, int predictionLen) {
        List<BigDecimal> yHat = estimateYHat(y, params, cycleLen, predictionLen);
        // wrap matrix
        yHat.remove(0);
        Matrix yHatMat = new Matrix(yHat.size(), 1);
        for (int i = 0; i < yHatMat.getRowNum(); i++) {
            yHatMat.setValue(i, 0, yHat.get(i));
        }
        return yHatMat;
    }

    private List<BigDecimal> estimateYHat(Matrix y, Matrix params, int cycleLen, int predictionLen) {
        BigDecimal alpha = params.getValue(0, 0);
        BigDecimal beta = params.getValue(0, 1);
        BigDecimal gama = params.getValue(0, 2);

        List<BigDecimal> s = new ArrayList<>();
        List<BigDecimal> b = new ArrayList<>();
        List<BigDecimal> c = new ArrayList<>();
        List<BigDecimal> yHat = new ArrayList<>();
        // Initialize s0
        BigDecimal sum = new BigDecimal(0.0);
        for (int i = 0; i < cycleLen; i++) {
            sum = sum.add(y.getValue(i, 0));
        }
        BigDecimal s0 = sum.multiply(new BigDecimal(1.0 / cycleLen));
        s.add(s0);
        // Initialize b0
        sum = new BigDecimal(0.0);
        for (int i = cycleLen; i < 2 * cycleLen; i++) {
            sum = sum.add(y.getValue(i, 0).subtract(y.getValue(i - cycleLen, 0)));
        }
        BigDecimal b0 = sum.multiply(new BigDecimal(1.0 / (cycleLen * cycleLen)));
        b.add(b0);
        // Initialize c
        for (int i = 0; i < cycleLen; i++) {
            BigDecimal ci = y.getValue(i, 0).subtract(s.get(0));
            c.add(ci);
        }
        // Initialize yHat
        BigDecimal y0 = s.get(0).add(b.get(0)).add(c.get(0));
        yHat.add(y0);
        // Estimate yHat
        for (int i = 0; i < y.getRowNum(); i++) {
            // estimate si
            BigDecimal sLeftTerm = alpha.multiply(y.getValue(i, 0).subtract(c.get(i)));
            BigDecimal sRightTerm = new BigDecimal(1.0 - alpha.doubleValue()).multiply(s.get(i).add(b.get(i)));
            s.add(sLeftTerm.add(sRightTerm));
            // estimate bi
            BigDecimal bLeftTerm = beta.multiply(s.get(i + 1).subtract(s.get(i)));
            BigDecimal bRightTerm = new BigDecimal(1.0 - beta.doubleValue()).multiply(b.get(i));
            b.add(bLeftTerm.add(bRightTerm));
            // estimate ci
            BigDecimal cLeftTerm = gama.multiply(y.getValue(i, 0).subtract(s.get(i)).subtract(b.get(i)));
            BigDecimal cRightTerm = new BigDecimal(1.0 - gama.doubleValue()).multiply(c.get(i));
            c.add(cLeftTerm.add(cRightTerm));
            // estimate yHat
            yHat.add(s.get(i + 1).add(b.get(i + 1)).add(c.get(i + 1)));
            // predict here
            if (i == y.getRowNum() - 1) {
                for (int j = 2; j <= predictionLen + 1; j++) {
                    BigDecimal lastTerm = new BigDecimal(j % predictionLen);
                    yHat.add(s.get(i + 1).add(b.get(i + 1).multiply(new BigDecimal(j))).add(c.get(i + 1)).add(lastTerm));
                }
            }
        }
        return yHat;
    }

    public class HoltWintersModelRmseTargetFunction implements TargetFunction {

        private int cycleLen;

        public HoltWintersModelRmseTargetFunction(int cycleLen) {
            this.cycleLen = cycleLen;
        }

        @Override
        public BigDecimal fun(Matrix params, Matrix args) {
            Matrix truthOutput = AlgebraUtil.getColumnVector(args, args.getColNum() - 1);
            return calcRmse(truthOutput, params);
        }

        private BigDecimal calcRmse(Matrix y, Matrix params) {
            List<BigDecimal> yHat = estimateYHat(y, params, cycleLen, 0);
            // Calculate RMSE
            BigDecimal sum = new BigDecimal(0.0);
            for (int i = 0; i < y.getRowNum(); i++) {
                sum = sum.add(y.getValue(i, 0).subtract(yHat.get(i)).pow(2));
            }
            BigDecimal rmse = new BigDecimal(Math.sqrt(sum.multiply(new BigDecimal(1.0 / y.getRowNum())).doubleValue()));
            return rmse;
        }

    }

}
